graphql_compiler/compiler/ir_lowering_cypher/ir_lowering.py

```diff
@@ -0,0 +1,185 @@
+# Copyright 2019-present Kensho Technologies, LLC.
+from functools import partial
+
+from ..blocks import CoerceType, Filter, Fold, MarkLocation, Recurse, Traverse
+from ..expressions import (
+    BinaryComposition, ContextField, FoldedContextField, LocalField, NullLiteral
+)
+from ..helpers import get_only_element_from_collection
+from ..ir_lowering_common.common import merge_consecutive_filter_clauses
+from ..ir_lowering_common.location_renaming import (
+    make_location_rewriter_visitor_fn, make_revisit_location_translations
+)
+from ..helpers import FoldScopeLocation
+
+
+##################################
+# Optimization / lowering passes #
+##################################
+
+def insert_explicit_type_bounds(ir_blocks, query_metadata_table, type_equivalence_hints=None):
+    """Add a CoerceType block after every Traverse/Fold/Recurse, to hint to the Cypher scheduler."""
+    # Cypher might not be aware of the fact that all our edges' endpoints are strictly typed,
+    # so we expose the implicit types of edges' endpoints explicitly, by adding CoerceType blocks.
```

Comment:
Is the cypher-generated graphql schema going to know about endpoint types? How?

Reply:
The same question will be relevant for the SPARQL backend.

Reply:
I think the schema generation process for these types of databases will need to be somewhat different compared to the existing ones. For example, it could be human-written, or it could be generated from a SHACL specification: https://www.w3.org/TR/shacl/